### PR TITLE
Multistart Solver Wrapper

### PR DESCRIPTION
**Pull request by sjkale**
Implemented Random Restarts for NLP's. Used High confidence stopping rule as following:

https://www.semanticscholar.org/paper/How-many-random-restarts-are-enough-Dick-Wong/55b248b398a03dc1ac9a65437f88b835554329e0

### CODE REVIEW

File: pyomo/solvers/plugins/solvers/multi.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+stopping_delta = .5
+
+class Multistart(OptSolver):
+    '''Solver wrapper that can check multiple starting points
+    '''
+    pyomo.util.plugin.alias('multistart',doc=textwrap.fill(textwrap.dedent(__doc__.strip())))
+
+    def __init__(self):
+        pass
+    def solve (self,model,**kwds):
```
Comment (qtothec) [inline pyomo/solvers/plugins/solvers/multi.py]: Documentation of keywords

File: pyomo/solvers/plugins/solvers/multi.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import copy
+from pyomo.core.base.var import Var
+from pyomo.core.kernel.numvalue import value
+from pyomo.core.base.objective import maximize
+from pyomo.environ import *
+import math
+logger = logging.getLogger('pyomo.solver')
+
+stopping_mass = .5
+stopping_delta = .5
```
Comment (qtothec) [inline pyomo/solvers/plugins/solvers/multi.py]: Documentation

File: pyomo/solvers/plugins/solvers/multi.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    def solve (self,model,**kwds):
+        strategy = kwds.pop('strategy','rand')
+        solver = kwds.pop('solver','ipopt')
+        iterations = kwds.pop('iterations',-1)
+        solver = SolverFactory(solver)
+        objectives = []
+        models = []
+        m = copy.deepcopy(model)
+        result = solver.solve(m)
+        if result.solver.status is SolverStatus.ok and result.solver.termination_condition is  TerminationCondition.optimal:
```
Comment (qtothec) [inline pyomo/solvers/plugins/solvers/multi.py]: line too long

File: pyomo/solvers/plugins/solvers/multi.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                lb = var.lb
+                ub = var.ub
+                # apply strategy to bounds/variable
+                strategies = {"rand": rand,
+                              "midpoint_guess_and_bound": midpoint_guess_and_bound,
+                              "rand_guess_and_bound": rand_guess_and_bound,
+                              "rand_distributed": rand_distributed
+                              }
+                var.value = strategies[strategy](val, lb, ub)
+    def num_one_occurrences(self,lst):
```
Comment (qtothec) [inline pyomo/solvers/plugins/solvers/multi.py]: What does this do?

File: pyomo/solvers/plugins/solvers/multi.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        dist = {}
+        for x in lst:
+            if x in dist:
+                dist[x] += 1
+            else:
+                dist[x] = 1
+        one_offs = [x for x in dist if dist[x] == 1]
+        return len(one_offs)
+
+    def should_stop(self,solutions):
```
Comment (qtothec) [inline pyomo/solvers/plugins/solvers/multi.py]: More documentation

**Comment by whart222** [comment]
In addition to inline documentation, this PR needs (a) Sphinx documentation and (b) tests that illustrate its use.

**Comment by qtothec** [comment]
@sjkale For sphinx documentation, take a look in https://github.com/Pyomo/pyomo/tree/master/doc/OnlineDocs. I believe that we're asking you for an *.rst file describing the functionality and usage.

**Comment by whart222** [comment]
@sjkale @qtothec Just to confirm, the Sphinx documentation is in *.rst files.  the pyomo/doc/OnlineDocs directory is where this documentation resides.  Perhaps in the solver subdirectory?

File: pyomo/solvers/tests/checks/test_multi.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+            m2 = build_model()
+            optsolver = SolverFactory('multistart')
+            optsolver.solve(m2,iterations = -1,HCS_param=(.99,.99))
+
+            self.assertTrue(value(m2.obj) >= value(m.obj))
+            del m2
+
+#simple non-convex model with many multiple minima
+def build_model():
```
Comment (blnicho) [inline pyomo/solvers/tests/checks/test_multi.py]: You should run these tests on a few different toy models with different features (Constraints, multiple variables, fixed variables, variables with no bounds, objectives to be minimized, etc.). In particular, make sure to include test cases that exercise any of the logic that you use in your solver class that checks for different model attributes.

**Comment by blnicho** [comment]
@sjkale any response to the comments in my review?

**Comment by blnicho** [comment]
@qtothec looks like there are some tests failing, could you take a look?

**Pull request merged**